python/tank_vendor/shotgun_authentication/interactive_authentication.py

```diff
@@ -0,0 +1,432 @@
+            """
+            QtCore.QObject.__init__(self)
+            self._res = None
+            self._exception = None
+            # Make sure that the invoker is bound to the main thread
+            self.moveToThread(QtGui.QApplication.instance().thread())
+
+        def __call__(self, fn, *args, **kwargs):
+            """
+            Asks the MainTheadInvoker to call a function with the provided parameters in the main
+            thread.
+            :param fn: Function to call in the main thread.
+            :param args: Array of arguments for the method.
+            :param kwargs: Dictionary of named arguments for the method.
+            :returns: The result from the function.
+            """
+            self._fn = lambda: fn(*args, **kwargs)
+            self._res = None
+
+            logger.debug("Sending ui request to main thread.")
+
+            QtCore.QMetaObject.invokeMethod(self, "_do_invoke", QtCore.Qt.BlockingQueuedConnection)
+
+            # If an exception has been thrown, rethrow it.
+            if self._exception:
+                raise self._exception
+            return self._res
+
+        @QtCore.Slot()
+        def _do_invoke(self):
+            """
+            Execute function and return result
+            """
+            try:
+                logger.debug("Invoking from main thread.")
+                self._res = self._fn()
+            except Exception, e:
+                self._exception = e
+
+    return MainThreadInvoker()
+
+
+class AuthenticationHandlerBase(object):
+    """
+    Base class for authentication requests. It handles locking reading cached credentials
+    on disk and writing newer credentials back. It also keeps track of any attempt to cancel
+    authentication.
+    """
+
+    def authenticate(self, host, login, http_proxy):
+        """
+        Does the actual authentication. Prompts the user and validates the credentials.
+        :param host Host to authenticate for.
+        :param login: User that needs authentication.
+        :param http_proxy: Proxy to connect to when authenticating.
+        :raises: TankAuthenticationError If the user cancels the authentication process,
+                 this exception will be thrown.
+        """
+        raise NotImplementedError
+
+    def _get_session_token(self, hostname, login, password, http_proxy):
+        """
+        Retrieves a session token for the given credentials.
+        :param hostname: The host to connect to.
+        :param login: The user to get a session for.
+        :param password: Password for the user.
+        :param http_proxy: Proxy to use. Can be None.
+        :returns: If the credentials were valid, returns a session token, otherwise returns None.
+        """
+        try:
+            return session_cache.generate_session_token(hostname, login, password, http_proxy)
+        except AuthenticationError:
+            return None
+
+    def raise_authentication_cancelled_error(self):
```

Comment:
At some point I was throwing an AuthenticationError with a custom message, so I refactored the code to a single method so the message was centralized for all places that needed it. Then later on I refactored that code to create a custom exception type, but didn't go through refactoring the method out. It's gone now.